### ISSUE TITLE
「分析完整性审计」：让 trace 不只回答"它做了什么"，还回答"这样做站得住吗"——数据变更 diff + 统计假设检查 + 分叉路径记录

### ISSUE THREAD

**Issue by HERRY423** [SuperJJ007/papermachine#8]
**可追溯 ≠ 可信。一个假设全违反、p 值挑出来的回归，trace 再完整也只是把错误记录得更清楚。**

## 背景 / 问题

PaperMachine 现在对 "trustworthy" 的定义是：**过程可见**（trace）+ **来源可查**（provenance）+ **数据可看**（transparency）。这三者解决的是「它到底做了什么」。但研究者、审稿人真正的问题是下一层：

> 它做的这些，**在统计上/方法上站得住吗？** 它是不是试了 5 种剔除方案、3 种模型，然后挑了一个 p < .05 的给我？

目前这层完全缺失，而且 README 自己也承认了一部分：

- "There is no variables panel yet"
- Roadmap: "A variable history view: shape changes of each dataset across cleaning steps"

我想把这个 roadmap 项目往前推一步，并把它和两个相邻能力合成一个**「分析完整性审计」**面板，因为三者共享同一份数据源——session log 里的 run 序列 + 持久 kernel 的状态。

### 具体缺口

**1. 数据变更没有 diff。** README 的例子是「审稿人问那 600 行是怎么被丢掉的」。现在你能找到那段代码，但你**需要自己读代码**才能知道是 600 行而不是 601 行，是 `dropna` 丢的还是 `query` 丢的。`get_science_state` 有 kernel 状态，但没有 run 前后的对比。

**2. 统计假设无检查。** 模型跑 `smf.ols(...)` 后直接报 coef 和 p 值。没有人（模型、产品、用户）检查残差正态性、异方差、多重共线性、样本量是否够、是否在同一 outcome 上做了多次检验。bundled 的 `statistical-analysis` skill 是 prompt 层的建议，**没有运行时保证**。

**3. 分叉路径（garden of forking paths）不可见。** Agent 在一个 turn 里可能 run 了 6 次：第 1 次 `dropna` 后 n=1200，第 3 次改成 `fillna(median)` 后 n=1800，第 5 次模型显著了，于是第 6 次画图。Trace 把 6 步都摊开了，但用户看到的是最终结论，**没有任何东西告诉用户"这个结论是第 3 条路"**。这恰好是 AI 分析工具最被诟病的 p-hacking-by-agent 风险，也是审稿人最想知道的。

## 竞品视角

| | 数据 diff | 假设检查 | 分叉路径记录 |
|---|---|---|---|
| Julius / ChatGPT ADA / Claude | ✗ | 靠 prompt 运气 | ✗ |
| JASP / jamovi | 部分（GUI 自带） | ✓ 但非 agent | ✗ |
| statcheck / p-checker | ✗ | 只查报告文本 | ✗ |
| **PaperMachine（提案后）** | ✓ 自动 | ✓ 运行时 | **✓ 唯一** |

这是产品从「AI 帮你做分析」升级到「AI 帮你做**经得起审的**分析」的关键一步，直接对应 README 的定位句 *"You supervise the AI"*——现在的 supervision 工具只有眼睛，我想给它一把尺子。

### 模块 A：数据变更 diff（先做，roadmap 已有）

在 kernel driver 里，每次 run 结束后对**本次 run 触碰过的 DataFrame / data.frame / tibble** 做一次廉价快照：

```json
{ "name": "df", "shape": [1800, 14], "columns": ["country", "year", ...],
  "dtypes": {...}, "naCount": {...}, "sha256OfIndex": "..." }
```

- Python：driver 在 run 前后遍历 `globals()` 中的 `pd.DataFrame`（限制大小；shape/columns/dtypes/isna().sum() 都是 O(n) 且不复制）
- R：`ls(globalenv())` + `inherits(x, "data.frame")`，同样只取 `dim` / `names` / `sapply(class)` / `colSums(is.na())`
- diff 结果作为新的 durable 事件 `science/variables-diffed`（扩展 `SessionEventMap`），只存元数据，不存内容——和现有 "Output text never enters a Science Session event" 原则一致
- UI：trace 的每个 step 徽章旁显示 `df: 2400 → 1800 rows (−600), +2 cols`；点开显示列级 diff（哪些列 NA 数变了、dtype 变了）。这就是那个 "variables panel"，而且是**带历史的**。

### 模块 B：运行时统计假设检查（"lint for analysis"）

`tools/post-execute` waterfall 里挂一个 checker，识别 run 结果中的常见模型对象并追加一次**只读**的诊断 run（用同一个 kernel epoch，不改变用户变量）：

| 检测到 | 自动跑 | 提示 |
|---|---|---|
| statsmodels OLS / R `lm` | Breusch-Pagan、Jarque-Bera / Shapiro、VIF、Cook's D、n/k 比 | ⚠ 异方差 p=.003，建议 HC3 |
| `ttest_ind` / `t.test` | Levene、各组 n、效应量 (Cohen's d) | ⚠ 未报效应量 |
| 同一 turn 内 ≥3 次对同一 outcome 的检验 | — | ⚠ 多重比较，未校正 |
| `sklearn` fit 在 `train_test_split` 之前做了 `fit_transform` | 静态 AST 检查 | ⚠ 疑似数据泄漏 |
| `lme4::lmer` | 收敛警告、singular fit | ⚠ singular，随机效应结构过复杂 |
| 任意 p 值 | — | 记录到分叉路径账本（模块 C） |

关键设计：

- **不打断**、**不改代码**、**不给模型打分**：只在 trace 的 step 上挂黄色徽章 + 一句话，用户点开看诊断 run 的代码和输出（它本身也是一个有完整 provenance 的 run）
- 用户可以一键把诊断结果**注入下一轮**（`agent.inject()`），让模型自己修
- 检查规则用 skill 同样的方式放在 `~/.papermachine/checks/`，用户/学科社区可以扩展（这也自然对接 roadmap 的 "Discipline environments"——社科环境配社科检查规则）

### 模块 C：分叉路径账本（Forking-paths ledger）

这是三者中最有差异化的，也几乎是免费的，因为数据都在 session log 里：

- 从 `science/run-*` 序列 + 模块 A 的 shape diff + 模块 B 抓到的检验结果，构建一棵树：**每次样本定义变化（n 变了）或模型规格变化 = 一个分支**
- 面板显示：本结论之前尝试了 4 条路径；每条路径的 n、模型、关键 p 值/效应量；哪条被采纳
- 导出为一段 **methods-ready 文字**（配合 bundled 的 `scientific-writing` skill）：

> "We initially excluded rows with missing GDP (n = 1,200); because this removed 33% of observations, we instead imputed by continent median (n = 1,800). Results under both specifications are reported in Supplementary Table S1."

这段话就是审稿人想看的，也是研究者最懒得写的。**把 agent 的试错过程变成 robustness check 的附录**，而不是需要藏起来的东西。

## 和架构的对接点

- 变量快照 / 诊断结果 → 新 `SessionEventMap` 条目，走 `dsh-science-session` strict fold
- 诊断 run → 复用 `startRun`，标记 `origin: 'audit'`，共享 kernel epoch（README 已保证 epoch 语义：同 epoch = 共享内存）
- checker → `tools/post-execute` waterfall listener
- 注入修复 → `agent.inject()`
- 面板 → `ConversationNodeDefinition` + keyed renderer（cookbook: adding-a-conversation-node）
- 规则扩展 → 与 `skill-filesystem` 相同的 nearest-scope-wins 解析

## 验收标准

- [ ] Gapminder 示例中 `dropna()` 后，step 徽章显示行数变化，点开可见列级 NA diff
- [ ] 跑 OLS 后 5 秒内出现假设检查徽章；诊断 run 在 trace 中可见且有 provenance
- [ ] 诊断 run 不改变用户 kernel 中任何变量（测试：前后 `globals()` 快照一致）
- [ ] 一个 turn 内切换两次样本定义后，分叉路径面板显示 ≥2 条路径并标出被采纳的一条
- [ ] 导出的 methods 段落包含所有分支的 n 与规格
- [ ] 所有检查可通过 Settings 关闭；关闭后无任何额外 run
- [ ] 性能：对 10^6 行 DataFrame 的快照开销 < 200 ms（只做 shape/dtype/NA count）

## 分期建议

- **MVP**：模块 A（variables diff）——它是 roadmap 已列的项，且是 B、C 的数据基础
- **v2**：模块 B 的前 3 条规则（OLS 诊断、t-test 效应量、多重比较计数）
- **v3**：模块 C + methods 文本导出 + 规则扩展机制

---

模块 A 我可以直接开工（driver 快照 + 事件 + 徽章 UI）。模块 B 的规则集希望 maintainer 或社区里有统计背景的人一起 review，避免做成"假警报机器"——宁可少而准。

**Comment by SuperJJ007** [comment]
感谢详细的提案，方向认可，会在后续版本中逐步加入；当前优先处理 Windows 安装相关的缺陷。
